d3a_interface/sim_results/kpi.py

```diff
@@ -145,11 +146,59 @@ def update_area_kpi(self, area_dict, core_stats):
         self._accumulate_energy_trace(core_stats)
 
 
+class SavingKPI:
+    def __init__(self):
+        self.producer_list = list()
+        self.consumer_list = list()
+        self.ess_list = list()
+        self.house_list = list()
```

Comment:
Again minor, but converting these to sets will help you avoid the `if_not_in_list_append` calls. Furthermore, judging from the ways that these lists are used, I would recommend not having 4 distinct lists, but 2: `producer_ess_list/set` , `consumer_ess_list/set`